design_notes/XDN05 - xlang Binary Interface.md

```diff
@@ -12,6 +12,233 @@ status: draft
+from IXlangObject. This interface adds additional core functionality needed for xlang.
+
+``` cpp
+enum ObjectInfoCategory
+{
+    StringRepresentation,
+    TypeName,
+    MemoryUsage
+};
+
+struct IXlangObject
+{
+    int32_t GetInfo(ObjectInfoCategory info_category, void** info);
+};
+```
+
+### GetInfo
+
+In WinRT, the IInspectable interface provides property get methods that retrieved information about
+the object needed in specific runtime scenarios. Over time, the list of information categories that
+might be useful at runtime grew. For example, knowing the memory footprint of a given WinRT object
+would allow for garbage collected languages like C# make better scheduling decisions. But since
+IInspectable is the base COM interface for every interface in WinRT, it's impossible to extend.
+
+For xlang, instead of individual property get methods, IXlangObject exposes a single GetObjectInfo
+method. The method takes an enum value indicating the object info being requested and a void** to
+hold the returned value. xlang objects are not required to support all object information
+categories. Similar to QueryInterface, an object can simply return an error code indicating that a
+requested information category is not supported by the object.
+
+Having a single GetInfo method has two primary benefits over the IInspectable approach. Because
+enums can be additively versioned over time, it is possible to add new ones over time. Since
+supporting a given information category is optional, adding new ones would not break existing
+objects. Furthermore, it reduces the size of virtual function table that every xlang object needs
+to support.
+
+Over time, we expect the list of object information categories to grow. Unlike COM interfaces, we
+expect the list of object information categories to be managed centrally by the xlang project.
+
+#### String Representation
+
+Most mainstream languages including
+[JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/toString),
+[Python](https://docs.python.org/3/library/functions.html#func-str),
+[.NET](https://docs.microsoft.com/en-us/dotnet/api/system.object.tostring?view=netframework-4.7.2),
+[Java](https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#toString()) and
+[Objective-C](https://developer.apple.com/documentation/objectivec/1418956-nsobject/1418746-description)
+have a mechanism to retrieve the string representation of an object. 
+
+The StringRepresentation information category retrieves the string representation of the object in
+question. The property type for StringRepresentation is an xlang_string.
+
+While all information categories are optional, it is highly recommended that all objects support
+StringRepresentation. By default, this method can return the type's name, however it can be
+implemented in whatever manner makes most sense for the underlying object.
+
+#### Type Information
```

Comment:
this section hints at a reflection facility, making xlang 'usable from dynamic languages'.  but type name alone is hardly sufficient.  what's the use case other than downcasting (to statically known types)?

Reply:
You're right, the name alone is not sufficient - you also need metadata. GetRuntimeClassName was added to enable JS to locate metadata for a given type on the fly. 

I don't follow your use case question. Can you rephrase?

Reply:
I've added text to indicate the name is not sufficient, that a dynamic language projection would need a mechanism to find (by name) and consume the metadata for this type, and that we need an XDN for this.